### PR TITLE
Router enhancements

### PR DESCRIPTION
**Pull request by mikemurray**
This update improves the router 404 handling. It used to redirect to a `/not-found` route. With this update, it will no longer redirect but instead display the `not-found` page on the current route.

This should bring the new React Router closer the functionality of the old router.

- Add React Router Switch component to App
- Add jsdoc to Router class
- Add a final, not-found route when no routes match
- Remove redirect to /not-found
- Cleanup code

### CODE REVIEW

File: imports/plugins/core/router/lib/router.js [hunk 1]
```diff
@@ -35,12 +34,41 @@ if (Meteor.isClient) {
 
 // Base router class (static)
```
Comment (machikoyasuda) [inline imports/plugins/core/router/lib/router.js]: If you could change the syntax here to jsdoc style: `/** Class representing a static base router */`, jsdoc will render the rest of the class.

File: imports/plugins/core/router/lib/router.js [hunk 2]
```diff
@@ -35,12 +34,41 @@ if (Meteor.isClient) {
...
+
+  /**
+   * Active classname for active routes
+   * @type {String}
+   */
   static activeClassName = "active";
 
+  /**
+   * Routes array
+   * @type {[Array]}
```
Comment (machikoyasuda) [inline imports/plugins/core/router/lib/router.js]: `{Array}` -- jsdoc can't parse the `[` and `]` inside `{` and `}`s

File: imports/plugins/core/router/lib/router.js [hunk 3]
```diff
@@ -49,61 +77,109 @@ class Router {
...
       }
     }
 
+    // Update route
     Router.go(current.route.name, current.params, queryParams);
   }
 
+  /**
+   * Watch path change. Is Reactive.
+   * @return {[type]} [description]
```
Comment (machikoyasuda) [inline imports/plugins/core/router/lib/router.js]: jsdoc can't parse the `[` and `]` inside `{` and `}`s - so, {description}`? what does this method return?

**Comment by mikemurray** [comment]
@machikoyasuda updated with your suggestions.

**Comment by aaronjudd** [comment]
badurl_and_2__reaction_reset_-y____reaction__node_

**Comment by mikemurray** [comment]
@aaronjudd The invalid tag is still a valid route because of the match `/tag/jdftfjshfgjsfg`; though not a tag, still matches route definition.

Think in that case the product grid has to determine what's valid, and display its own custom 404 page like the PDP does.

---

Otherwise, we need a routing table (of sorts) with all possible routes to match against to determine when one truly doesn't exist.

**Comment by mikemurray** [comment]
Confirmed it's happening in master as well. Going to make a new PR with a not found page / view for the product grid.

screen shot 2017-08-08 at 2 14 56 pm

**Comment by mikemurray** [comment]
PR for product grid not found view on broken tag #2656, currently WIP.

**Pull request merged**